NHM & Cybershake Examples

claudio525 opened the pull request:
- Adds functionality to support for the computation of GMM-based NHM hazard for both fault-based and distributed hazard.
- Adds functionality to support computation of Cybershake based hazard
- Adds NHM hazard files

Also includes example scripts for computing GMM-based NHM hazard and Cybershake hazard and plotting the hazard curve.

claudio525 committed:
Added functionaility & example for computing simulation-based (fault) seismic hazard for the 2010 NSHM.

claudio525 committed:
GMM-based fault hazard calculation for 2010 NSHM

claudio525 committed:
Added DS hazard computation for GMM-based hazard (for the 2010 NSHM).

claudio525 committed:
Merge branch 'main' into hazard_example

claudio525 committed:
Update fault creation from NHM fault data. Tidy up.

claudio525 committed:
Tidy up.

joelridden commented on examples/NSHM2010/gmm_hazard.py:
```diff
@@ -0,0 +1,121 @@
+from qcore import coordinates as coords
+
+# Periods to compute hazard for
+PERIODS = [
```
Do we want these periods to be the same as what we do for simulations / observed (aligned with NGA-West3)?

sungeunbae replied:
This is just an example code, so I'm ok with keeping it simple like this. Perhaps leaving a comment containing a link to the list of periods (eg. IM_calc) should be enough.

claudio525 replied:
Yeah its just an example, so up to the user to change it. @sungeunbae not sure what you mean by list of periods in IM calc? In theory the user can specify whatever periods they want as long as the are supported by the GMMs?

sungeunbae replied:
@claudio525 I meant a link showing standard periods used by IM calculation.

https://github.com/ucgmsim/IM_calculation/blob/dea8c12a28152298b06cebcbc47981e20727870d/IM/im_calculation.py#L10

joelridden commented on seismic_hazard_analysis/nshm_2010/gmm_hazard.py:
```diff
@@ -0,0 +1,267 @@
+        scenario_section_ids,
+        plane_nztm_coords,
+        segment_section_ids,
+        # site_nztm[0, [1, 0, 2]],
```
Test code I think?

joelridden commented on seismic_hazard_analysis/site_source.py:
```diff
@@ -0,0 +1,977 @@
+        shape: [n_segments]
+    """
+    # Make matrix of all unique trace points
+    # unique_trace_points = np.unique(
```
Is it worth still having the commented code out that requires the different solution as numba doesnt support it?

claudio525 replied:
I think so (at least in my opinion), its a) much easier to read, and b) hopefully at some stage numba will support the axis parameter.

joelridden commented on seismic_hazard_analysis/site_source.py:
```diff
@@ -0,0 +1,977 @@
+    # Assume the strike is defined from point 0 -> point 2
+    # This did not hold for the new NSHM rupture definitions hence the logic below
+    # However, the below does not work for faults with dip=90 degrees
+    # TODO: Fix this!!
```
Is this a later PR problem?

joelridden commented on seismic_hazard_analysis/utils.py:
```diff
@@ -168,3 +168,58 @@ def is_pd(B: np.ndarray[float]):
         return False
+
+
+def get_min_max_levels_for_im(im: str):
```
Some of this might need to change if dealing with new IM outputs due to changed units for CAV etc.
Might be completely fine but good to double check

claudio525 replied:
This is for hazard, doesn't have anything to with IM calc? Or are you just meaning for consistency?

joelridden replied:
I guess for consistency, wasnt sure if this function was going to use IM_calc outputs to generate hazard either

sungeunbae commented on seismic_hazard_analysis/site_source.py:
```diff
@@ -0,0 +1,977 @@
+
+    Returns
+    -------
+    rjb: float
```
in km, I believe?  It can be obvious, but there are few other places where the unit used could be documented explicitly.

claudio525 committed:
PR comments + minor fix

AndrewRidden-Harper commented on seismic_hazard_analysis/nshm_2010/gmm_hazard.py:
```diff
@@ -0,0 +1,266 @@
+        OpenQuake GMM wrapper.
+    gmm_mapping: dict
+        Specifies the GMM to use for each tectonic type
+    pSA_periods:
```
Missing a description of pSA_periods

AndrewRidden-Harper commented on seismic_hazard_analysis/nshm_2010/gmm_hazard.py:
```diff
@@ -0,0 +1,266 @@
+        )
+        / 1000
+    )
+    # Todo: Not sure about this. This is just what we did in the past.
```
I'm happy for this comment to stay as it is, but if you want, more details could be added

claudio525 replied:
Good catch, will have another think about this one! Setting it to 0 seems wrong..

AndrewRidden-Harper commented on seismic_hazard_analysis/site_source.py:
```diff
@@ -0,0 +1,953 @@
+    ry_origins: array of floats
+        shape: [2, n_segments]
+        The origin used for each Ry calculation
+    segment_section_ids: array of ints
```
Is segment_section_ids returned?

AndrewRidden-Harper commented on seismic_hazard_analysis/site_source.py:
```diff
@@ -0,0 +1,953 @@
+    """
+    Computes the Rjb, Rrup, Rx, and Ry distance
+    for each scenario. For full details off the arguments
+    see compute_single_scenario_distances
```
arguments aren't described in the docstring but they may not really be neccesasary

claudio525 replied:
All the arguments are just passed through to `compute_single_scenario_distances` which describes the arguments, so don't think its necessary to add them here again.

AndrewRidden-Harper reviewed (commented):
a few minor comments

claudio525 committed:
PR comments

claudio525 committed:
Use Rjb for Rx and Ry for DS instead of 0

claudio525 commented:
@lispandfound I had a look at the site_source.py module, and it does make use of numba. I'm pretty keen to leave this as is, since it will allow us to perform fast site-source calculation for cases where the number of sources is very large. Not needed for NHM2010, but this package isn't just specifically for NHM2010.

Mor-Abs committed:
Update simulation_hazard.py

The following line failed when working with versions 20.4 and 20.5. 
            else:
                assert np.all(cur_stations == cur_im_df.index)
                assert np.all(cur_IMs == cur_im_df.columns[1:])


To address the issue that not all CSV files might share the same stations (or IMs), I modified the code so that it includes all data into cur_im_array and removes stations and IMs with missing data.

lispandfound commented on seismic_hazard_analysis/nshm_2010/simulation_hazard.py:
```diff
@@ -0,0 +1,153 @@
+        )
+        fault_im_dict[cur_fault] = cur_im_array
+
+    return fault_im_dict
```
The `xarray` approach to this would be an xarray dataset, **especially** because these are indexed on the same coordinates.

lispandfound commented on seismic_hazard_analysis/nshm_2010/simulation_hazard.py:
```diff
@@ -0,0 +1,153 @@
+
+        cur_results.append(cur_df)
+
+    return pd.concat(cur_results, axis=0)
```
Having the input as an xarray dataset makes this a one-liner and you could probably remove this function.

lispandfound commented on seismic_hazard_analysis/nshm_2010/utils.py:
```diff
@@ -0,0 +1,202 @@
+        )
+        planes.append(plane)
+
+    return sources.Fault(planes)
```
There should be a `Fault` class method to create a fault from a trace.

lispandfound commented on seismic_hazard_analysis/utils.py:
```diff
@@ -168,3 +168,58 @@ def is_pd(B: np.ndarray[float]):
+    elif im.upper() == "MMI":
+        return 1.0, 12.0
+    else:
+        raise ValueError("Invalid IM")
```
```suggestion
    match im.upper():
        case _ if im.startswith("pSA"):
            period = float(im.rsplit("_", 1)[-1])
            periods = np.array([0.5, 1.0, 3.0, 5.0, 10.0])
            bounds = [(0.005, 10.0), (0.005, 7.5), (0.0005, 5.0), (0.0005, 4.0), (0.0005, 3.0)]
            idx = np.searchsorted(period, periods)
            return bounds[idx]
        case "PGA":
            return 0.0001, 10.0
        case "PGV":
            return 1.0, 400.0
        case "CAV":
            return 0.0001 * 980, 20.0 * 980.0
        case "AI":
            return 0.01, 1000.0
        case "DS575" | "DS595":
            return 1.0, 400.0
        case "MMI":
            return 1.0, 12.0 
        case _:
            raise ValueError("Invalid IM")
```

lispandfound replied:
Personally, I think a match statement makes this logic less repetitive.

lispandfound commented on tests/test_hazard.py:
```diff
@@ -77,8 +77,8 @@ def test_non_parametric_gm_prob():
-    assert result.loc["rupture_1"] == 0.5
-    assert result.loc["rupture_2"] == 2 / 3
+    assert np.isclose(result.loc["rupture_1", 2], 0.5)
+    assert np.isclose(result.loc["rupture_2", 2], 2 / 3)
```
```suggestion
    assert result.loc["rupture_1", 2] == pytest.approx(0.5)
    assert result.loc["rupture_2", 2] == pytest.approx(2 / 3)
```
Gives better output when tests fail.

lispandfound commented on seismic_hazard_analysis/site_source.py:
```diff
@@ -0,0 +1,950 @@
+                True
+                if id
+                in segment_section_ids[unique_section_id_ind[section_strike_flip_mask]]
+                else False
```
```suggestion
                id in segment_section_ids[unique_section_id_ind[section_strike_flip_mask]]
```

lispandfound commented on seismic_hazard_analysis/site_source.py:
```diff
@@ -0,0 +1,950 @@
+    n /= np.linalg.norm(n, axis=0)
+    v3 = segment_nztm_coords[3, :, :] - segment_nztm_coords[0, :, :]
+    v3 /= np.linalg.norm(v3, axis=0)
+    if np.any(np.einsum("ij,ij->j", n, v3) > 1e-2):
```
I would extract that einsum out and give it a name like `column_wise_dot_product` so that future readers can understand what this actually doing.

lispandfound commented on seismic_hazard_analysis/site_source.py:
```diff
@@ -0,0 +1,950 @@
+
+
+@nb.njit(cache=True)
+def f(A: np.ndarray, B: np.ndarray, t: float):
+    """Helper function for computing the minimum distance to a line segment"""
+    return (1 - t) * A + t * B
```
```suggestion
def interp(A: np.ndarray, B: np.ndarray, t: float):
    """Numba alternative to `np.interp`."""
    return (1 - t) * A + t * B
```

lispandfound commented on seismic_hazard_analysis/site_source.py:
```diff
@@ -0,0 +1,950 @@
+        The section id for each segment
+    segment_rjb: array of floats
+        Rjb distance for each segment
+    segment_rrup
+        Rrup distance for each segment
+    segment_rx
+        Rx distance for each segment
+    segment_ry
+        Ry distance for each segment
+    segment_ry_origin
+        The origin used for the
```
```suggestion
    segment_rrup : array of floats
        Rrup distance for each segment
    segment_rx : array of floats
        Rx distance for each segment
    segment_ry : array of floats
        Ry distance for each segment
    segment_ry_origin : array of floats
        The origin used for the
```

claudio525 committed:
Merge pull request #3 from ucgmsim/Mor-Abs-patch-1

Update simulation_hazard.py

claudio525 committed:
PR comments

claudio525 committed:
Merge branch 'hazard_example' of github.com:ucgmsim/seismic_hazard_analyis into hazard_example

claudio525 committed:
Fixes and tidy up

lispandfound commented:
@claudio525 could you add a pytest action for this repo?

claudio525 commented:
@lispandfound The next PR adds pytest based benchmark tests, https://github.com/ucgmsim/seismic_hazard_analyis/pull/4
